### PR TITLE
Improve Span.Count

### PR DESCRIPTION
**Pull request by MihaZupan**
We're currently calculating the number of remaining elements after the vectorized loop and falling back to the scalar path if we have less than half a vector's worth of data remaining.

It's simpler and faster if we just unconditionally process the last vector.

```c#
public class CountBenchmarks<T> where T : struct, INumber<T>
{
    private T[][] _numbers;

    [GlobalSetup]
    public void Setup()
    {
        _numbers = new T[10_000][];

        var rng = new Random(42);

        for (int i = 0; i < _numbers.Length; i++)
        {
            _numbers[i] = new T[rng.Next(0, 256 / Unsafe.SizeOf<T>())];
            rng.NextBytes(MemoryMarshal.AsBytes(_numbers[i].AsSpan()));
        }
    }

    [Benchmark]
    public int Count()
    {
        int count = 0;
        foreach (T[] numbers in _numbers) count += numbers.AsSpan().Count(T.CreateChecked(42));
        return count;
    }
}
```

| Method | Toolchain | Type  | Mean      | Error    | Ratio |
|------- |---------- |------ |----------:|---------:|------:|
| Count  | main      | byte  |  175.4 us |  0.82 us |  1.00 |
| Count  | pr        | byte  |  100.1 us |  0.25 us |  0.57 |
|        |           |       |           |          |       |
| Count  | main      | short |  147.6 us |  0.24 us |  1.00 |
| Count  | pr        | short |  102.7 us |  0.17 us |  0.70 |
|        |           |       |           |          |       |
| Count  | main      | int   | 143.01 us | 0.277 us |  1.00 |
| Count  | pr        | int   |  98.10 us | 0.262 us |  0.69 |
|        |           |       |           |          |       |
| Count  | main      | long  | 134.13 us | 2.157 us |  1.00 |
| Count  | pr        | long  |  99.87 us | 0.618 us |  0.74 |

@gfoidl this goes against your numbers in #82687, presumably because you were testing with a single `Length` at a time and all the branches were predictable?

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 1]
```diff
@@ -3786,92 +3786,59 @@ internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpa
                 {
                     Vector512<T> targetVector = Vector512.Create(value);
                     ref T oneVectorAwayFromEnd = ref Unsafe.Subtract(ref end, Vector512<T>.Count);
-                    do
+                    while (Unsafe.IsAddressLessThan(ref current, ref oneVectorAwayFromEnd))
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3789]: Why the change from a do...while to a while? Does it help? We just validated there's enough room, so we shouldn't need the starting check, right?

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 2]
```diff
@@ -3786,92 +3786,59 @@ internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpa
...
                         current = ref Unsafe.Add(ref current, Vector256<T>.Count);
                     }
-                    while (!Unsafe.IsAddressGreaterThan(ref current, ref oneVectorAwayFromEnd));
 
-                    // If there are just a few elements remaining, then processing these elements by the scalar loop
-                    // is cheaper than doing bitmask + popcount on the full last vector. To avoid complicated type
-                    // based checks, other remainder-count based logic to determine the correct cut-off, for simplicity
-                    // a half-vector size is chosen (based on benchmarks).
-                    uint remaining = (uint)Unsafe.ByteOffset(ref current, ref end) / (uint)Unsafe.SizeOf<T>();
-                    if (remaining > Vector256<T>.Count / 2)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3829]: Are there other APIs where we've done a similar guard, or just with Count?

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs (hunk 1 above)
Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs]: Correctness-wise, it doesn't matter, the overlapped block can handle any pattern.

Theoretically the `while (less than)`, we can skip the main loop iteration if the input is exactly one `VectorXYZ<T>` long. But it does mean an extra check for all the other cases. Perf-wise it makes practically no difference in this case.

The `do..while(less than)` saves a dozen bytes of assembly for the removed branches, so I went with that.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3829]: I don't remember seeing a similar pattern where we'd revert to the scalar loop depending on how much data remains. I'm not immediately seeing anything too similar on Grep.app.

The closest thing that comes to mind is this logic in ProbabilisticMap
https://github.com/dotnet/runtime/blob/24b59d85541c72fd55b8d5d0bc9adb84619d6c4f/src/libraries/System.Private.CoreLib/src/System/SearchValues/ProbabilisticMap.cs#L528-L540
where we switch between a final iteration of Vector128 (preferred) vs. a final iteration of Vector256.
But I did that there because it's possible the already-handled (overlapping) elements were false-positive matches we had to rule out one-by-one in the verification step, so the potential overhead of having to recheck them is a lot higher than here in `Count` where it's practically free.

**Comment by gfoidl** [comment]
> this goes against your numbers in #82687, presumably because you were testing with a single `Length` at a time and all the branches were predictable?

Probably. At least your benchmark (w/ different length) seems better. So a nice improvement.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs (hunk 2 above)
Comment (gfoidl) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3829]: > Are there other APIs where we've done a similar guard, or just with Count?

I also think this one was the only place where we did this.

**Review by stephentoub — APPROVED** [review]
